crates/gossip/src/burst.rs

```diff
@@ -0,0 +1,215 @@
+use crate::gossip::DropAbortHandle;
+use crate::K2GossipConfig;
+use kitsune2_api::{Timestamp, Url, UNIX_TIMESTAMP};
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::sync::{Arc, Mutex};
+use std::time::Duration;
+
+#[derive(Clone)]
+pub(crate) struct AcceptBurstTracker {
+    config: Arc<K2GossipConfig>,
+    state: Arc<Mutex<HashMap<Url, Vec<Timestamp>>>>,
+    _task: Arc<DropAbortHandle>,
+}
+
+impl Debug for AcceptBurstTracker {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AcceptBurstTracker")
+            .field("state", &self.state)
+            .finish()
+    }
+}
+
+impl AcceptBurstTracker {
+    pub(crate) fn new(config: Arc<K2GossipConfig>) -> Self {
+        let state: Arc<Mutex<HashMap<Url, Vec<Timestamp>>>> =
+            Default::default();
+
+        // A task is needed to clear out old timestamps. If we only clear them when checking a
+        // new accept, then we would never clear timestamps for peers that are no longer sending
+        // us accepts.
+        let task_config = config.clone();
+        let state_weak = Arc::downgrade(&state);
+        let handle = tokio::task::spawn(async move {
+            loop {
+                tokio::time::sleep(Duration::from_secs(5 * 60 * 1000)).await;
+
+                if let Some(state) = state_weak.upgrade() {
+                    Self::cleanup(&state, &task_config);
+                } else {
+                    tracing::debug!("AcceptBurstTracker task exiting");
+                    break;
+                }
+            }
+        });
+
+        Self {
+            config,
+            state,
+            _task: Arc::new(DropAbortHandle {
+                name: "AcceptBurstTracker".to_string(),
+                handle: handle.abort_handle(),
+            }),
+        }
+    }
+
+    pub(crate) fn check_accept(
+        &self,
+        peer_url: &Url,
+        timestamp: Timestamp,
+    ) -> bool {
+        let max_burst = Self::max_burst(&self.config);
```

Comment:
Why are you re-computing these values with every accept?  Is it because the K2 config is allowed to be modified during runtime?

Reply:
Yes, because config can be changed. In theory gossip would cope reasonably well with config being changed but we haven't actually tried that out yet, just left the possibility open